File: Sources/RMBTHistoryResultPercentView.m
//
//  RMBTHistoryResultPercentView.m
//  RMBT
//
//  Created by Sergey Glushchenko on 03.09.2020.
//  Copyright © 2020 appscape gmbh. All rights reserved.
//

#import "RMBTHistoryResultPercentView.h"

@implementation RMBTHistoryResultPercentView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    [self setBackgroundColor:[UIColor clearColor]];
    return self;
}

- (void)drawRect:(CGRect)rect {
    UIImage *image = [UIImage imageNamed:@"traffic_lights_template"];

    CGFloat pointSize = 7;
    NSInteger countPointsHorizontal = self.bounds.size.width / pointSize;
    NSInteger countPointsVertical = 2;
    
    NSInteger countFillPointsHorizontal = (self.bounds.size.width * self.percents) / pointSize;

    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextTranslateCTM(context, 0, self.bounds.size.height);
    CGContextScaleCTM(context, 1.0, -1.0);
    
    //Draw mask
    for (NSInteger i = 0; i < countPointsHorizontal; i++) {
        for (NSInteger j = 0; j < countPointsVertical; j++) {
            CGFloat x = i * pointSize;
            CGFloat y = j * pointSize;
            
            CGRect imageRect = CGRectMake(x, y, pointSize, pointSize);
            CGContextDrawImage(context, imageRect, [image CGImage]);
        }
    }
    
    //Create mask
    CGImageRef alphaMask = CGBitmapContextCreateImage(context);
    
    //Append mask
    CGContextClipToMask(context, self.bounds, alphaMask);
    [self.filledColor setFill];
    
    //Draw filled area
    CGRect fillRect = CGRectMake(0, 0, countFillPointsHorizontal * pointSize, self.bounds.size.height);
    CGContextFillRect(context, fillRect);
    
    //Draw unfilled area
    [self.unfilledColor setFill];
    CGRect unfillRect = CGRectMake(countFillPointsHorizontal * pointSize, 0, self.bounds.size.width, self.bounds.size.height);
    CGContextFillRect(context, unfillRect);
}

@end

File: Sources/History/Views/RMBTHistoryResultPercentView.swift
//
//  RMBTHistoryResultPercentView.swift
//  RMBT
//
//  Created by Sergey Glushchenko on 17.09.2021.
//  Copyright © 2021 appscape gmbh. All rights reserved.
//

import UIKit

@objc final class RMBTHistoryResultPercentView: UIView {

    @objc public var percents: CGFloat = 0.0 {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    @objc public var filledColor: UIColor = UIColor.white
    @objc public var unfilledColor: UIColor = UIColor.white.withAlphaComponent(0.3)
    @objc public var templateImage: UIImage? = UIImage(named: "traffic_lights_template")
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.clear
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func draw(_ rect: CGRect) {
        guard let image = self.templateImage,
              let context = UIGraphicsGetCurrentContext()
        else {
            return
        }
        
        let pointSize = image.size.height
        let countPointsHorizontal: Int = Int(self.bounds.size.width / pointSize)
        let countPointsVertical = 2
        
        let countFillPointsHorizontal = (self.bounds.size.width * self.percents) / pointSize
        
        context.translateBy(x: 0, y: self.bounds.size.height)
        context.scaleBy(x: 1.0, y: -1.0)
        context.setBlendMode(.copy)
        
        //Draw mask
        guard let cgImage = image.cgImage else { return }
        for i in 0..<countPointsHorizontal {
            for j in 0..<countPointsVertical {
                let x = CGFloat(i) * pointSize
                let y = CGFloat(j) * pointSize
                
                let imageRect = CGRect(x: x, y: y, width: pointSize, height: pointSize)
                context.draw(cgImage, in: imageRect)
            }
        }
        
        //Create mask
        guard let alphaMask = context.makeImage() else { return }
        
        //Append mask
        context.clip(to: self.bounds, mask: alphaMask)
        self.filledColor.setFill()
        
        //Draw filled area
        let fillRect = CGRect(x: 0, y: 0, width: countFillPointsHorizontal * pointSize, height: self.bounds.size.height)
        context.fill(fillRect)
        
        //Draw unfilled area
        self.unfilledColor.setFill()

        let unfillRect = CGRect(x: countFillPointsHorizontal * pointSize, y: 0, width: self.bounds.size.width, height: self.bounds.size.height)
        context.fill(unfillRect)
    }
}
